utils/publish-rpm.sh

```diff
@@ -1,138 +1,163 @@
-    ./ossutil64 cp -r oss://apisix-repo/packages/centos/7/x86_64 ./ --config-file=/tmp/ossutilconfig
-
-    # rebuild repo
-    cp ./${TARGET_APP}-${TAG_VERSION}-0.el7.x86_64.rpm ./x86_64
-    cd ./x86_64
-
-    sudo apt-get update
-    sudo apt install createrepo -y
-    createrepo .
-    cd ../
+func_oss_utils_credential_init() {
+    # ${1} - OSS endpoint
+    # ${2} - ACCESS_KEY_ID
+    # ${3} - ACCESS_KEY_SECRET
+    cat > "$(eval echo ~${USER})/.ossutilconfig" <<_EOC_
+[Credentials]
+language=EN
+endpoint=${1}
+accessKeyID=${2}
+accessKeySecret=${3}
+_EOC_
 }
 
-
-sign_repo_metadata() {
-    rm ./x86_64/repodata/repomd.xml.asc
-    gpg --batch --pinentry-mode loopback --passphrase-file /tmp/rpm-gpg-publish.passphrase --detach-sign --armor ./x86_64/repodata/repomd.xml
-
-    out=$(gpg --verify x86_64/repodata/repomd.xml.asc 2>&1)
-    if ! echo "$out" | grep -iq 'Good signature'; then
-        echo "failed: check rpm metadata signatures"
-        exit 1
-    fi
+# =======================================
+# OSS repo extension
+# =======================================
+func_repo_init() {
+    # ${1} - repo workbench path
+    mkdir -p "${1}"/centos/{7,8}/x86_64
 }
 
-
-upload_new_repo() {
-    # rm origin repo and upload new repo
-    ./ossutil64 rm -r -f oss://apisix-repo/packages/centos/7/x86_64 --config-file=/tmp/ossutilconfig
-    ./ossutil64 cp -r ./x86_64 oss://apisix-repo/packages/centos/7/x86_64 --config-file=/tmp/ossutilconfig
+func_repo_clone() {
+    # ${1} - bucket name
+    # ${2} - OSS path
+    # ${3} - target path
+    ossutil64 cp -r -f "oss://${1}/packages/${2}" "${3}"
 }
 
+func_repo_backup() {
+    # ${1} - bucket name
+    # ${2} - OSS path
+    # ${3} - backup tag
+    ossutil64 cp -r "oss://${1}/packages/${2}" "oss://${1}/packages/backup/${2}_${3}"
+}
 
-check_down_load_rpm() {
-    mkdir temp && cd temp
-    wget https://apisix-repo.oss-cn-shenzhen.aliyuncs.com/packages/centos/7/x86_64/${TARGET_APP}-${TAG_VERSION}-0.el7.x86_64.rpm
-    if [ ! -f ${TARGET_APP}-${TAG_VERSION}-0.el7.x86_64.rpm ]; then
-        echo "failed: download new ${TARGET_APP} rpm package"
-        exit 1
-    fi
-    cd ../
+func_repo_backup_remove() {
+    # ${1} - bucket name
+    # ${2} - OSS path
+    # ${3} - backup tag
+    ossutil64 rm -r -f "oss://${1}/packages/backup/${2}_${3}"
 }
 
+func_repo_repodata_rebuild() {
+    # ${1} - repo parent path
+    find "${1}" -type d -name "x86_64" \
+        -exec echo "createrepo for: {}" \; \
+        -exec rm -rf {}/repodata \; \
+        -exec createrepo {} \;
+}
 
-rm_backup_repo() {
-    ./ossutil64 rm -r -f oss://apisix-repo/packages/backup/centos/7/x86_64_${DATE_TAG} --config-file=/tmp/ossutilconfig
+func_repo_repodata_sign() {
+    # ${1} - repo parent path
+    find "${1}" -type f -name "*repomd.xml" \
+        -exec echo "sign repodata for: {}" \; \
+        -exec gpg --batch --pinentry-mode loopback --passphrase-file "${VAR_GPG_PASSPHRASE}" --detach-sign --armor {} \;
```

Comment:
Will this operation overwrite an existing `repomd.xml.asc`?

ref: https://github.com/api7/apisix-build-tools/blob/2b7b880e52f880ce48ed5a75bd9f12a6901665b0/utils/publish-rpm.sh#L83